headeredevent.go

```diff
@@ -0,0 +1,84 @@
+package gomatrixserverlib
+
+import (
+	"encoding/json"
+	"reflect"
+
+	"github.com/tidwall/sjson"
+)
+
+// HeaderedEventHeader contains header fields for an event that contains
+// additional metadata, e.g. room version. IMPORTANT NOTE: All fields in
+// this struct must have a "json:" name tag or otherwise the reflection
+// code for marshalling and unmarshalling headered events will not work.
+type EventHeader struct {
+	RoomVersion RoomVersion `json:"room_version"`
+}
+
+// HeaderedEvent is a wrapper around an Event that contains information
+// about the room version. All header fields will be added into the event
+// when marshalling into JSON and will be separated out when unmarshalling.
+type HeaderedEvent struct {
+	EventHeader
```

Comment:
Naming and JSON marshalling: This would read better if the field was called `Header`, rather than struct embedding like this. Also, how does this work for cases where we accidentally use the same JSON tag as an actual event value? The answer is, it panics.

This might be fine as long as we document it, as the utility of `headeredEvent.RoomVersion` may outweigh the fail-fast panic and the longness of `headeredEvent.Header.RoomVersion` everywhere.

Reply:
https://play.golang.org/p/ijjVmbVHACe

Reply:
So the main reason I chose `EventHeader` was because the entire of gomatrixserverlib only implements the `gomatrixserverlib` package, not subpackages. It makes it a bit clearer in `godoc` and friends. 

As for `struct field Bar repeats json tag` and duplicate struct tags, that's a documented Go case but I'm happy to add some more notes on it.

Reply:
It's worth being explicit: not everyone who contributes will be knowledgable about edge cases in Go. And you can still have `type EventHeader struct` but have the field be called `Header` you know.

Reply:
Indeed, although naming the field `Header` specifically stops us from getting top-level embedding. I'll update the comments.

Reply:
What's wrong with having the header a level further down?

Reply:
From a struct perspective, nothing at all. From a marshalling perspective, it doesn't make things much easier for us since we still need to handle that marshalling separately or otherwise Go just defaults to the `Event` marshaller (since it overloads the top-level) and it misses out the `EventHeader` fields altogether.

Reply:
Ohh, I see now. Yuck. I still think it would be clearer to have them namespaced rather than sitting together at the same level.